scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -279,9 +279,107 @@
   (-> (d/entity db [:instance/task-id task-id])
       :db/id))
 
+(histograms/defhistogram [cook-mesos scheduler progress-aggregator-pending-states])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-message-rate])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-drop-rate])
+(counters/defcounter [cook-mesos scheduler progress-aggregator-drop-count])
+
+(defn progress-update-aggregator
+  "Launches a long running go block that triggers publishing the latest aggregated instance-id->progress-state
+   whenever there is a read on the progress-state-chan.
+   It drops messages if the progress-aggregator-chan queue is larger than pending-progress-threshold or the aggregated
+   state has more than pending-progress-threshold different entries.
+   It returns the progress-aggregator-chan which can be used to send progress-state messages to the aggregator."
+  [pending-progress-threshold progress-state-chan]
+  (log/info "Starting progress update aggregator")
+  (let [progress-aggregator-chan (async/chan (async/sliding-buffer pending-progress-threshold))
+        progress-aggregator-fn (fn progress-aggregator-fn [instance-id->progress-state data]
+                                 (let [{:keys [instance-id response-chan]} data]
```

Comment:
I assume this response-chan is only here for testing. Please remove it, and then move this function into the namespace and test it explicitly. Given the tests on xform-pipe and the tests of this function, you can have simpler tests for the aggregator.